File: api/types/workout.go
package types

import (
	"context"
	"time"
)

type WorkoutStore interface {
	CreateWorkout(ctx context.Context, workout Workout, muscleIDS []int) error
	IsWorkoutDuplicated(name string) (v bool, err error)
	FilterWorkouts(filter WorkoutFilter, order WorkoutOrder, pagination Pagination) (workouts []Workout, err error)
	CountWorkouts(filter WorkoutFilter) (count int, err error)
	FindWorkout(id int) (workout *Workout, err error)
	UpdateWorkout(id int, workout Workout, muscleIDS []int) error
	DeleteWorkout(id int) error
}

type Workout struct {
	ID           int       `json:"id"`
	Name         string    `json:"name"`
	Description  string    `json:"description"`
	UserID       int       `json:"userId"`
	DifficultyID int       `json:"difficultyId"`
	CreatedAt    time.Time `json:"createdAt"`
}

type WorkoutWithRelations struct {
	ID          int        `json:"id"`
	Name        string     `json:"name"`
	Description string     `json:"description"`
	UserID      int        `json:"userId"`
	Muscles     []Muscle   `json:"muscles"`
	Difficulty  Difficulty `json:"difficulty"`
	CreatedAt   time.Time  `json:"createdAt"`
}

type WorkoutFilter struct {
	Name         string
	MuscleIDs    []int
	DifficultyID int
}

type WorkoutOrder struct {
	Name      string
	CreatedAt string
}

type QueryWorkoutsDTO struct {
	Filter FilterWorkoutsDTO `json:"filter"`
	Order  OrderWorkoutsDTO  `json:"order"`
}

type FilterWorkoutsDTO struct {
	Name         string `json:"name"`
	MuscleIDs    []int  `json:"muscleIds"`
	DifficultyID int    `json:"difficultyId"`
}

type OrderWorkoutsDTO struct {
	Name      string
	CreatedAt string
}

type CreateWorkoutDTO struct {
	Name         string `json:"name" validate:"required"`
	Description  string `json:"description"`
	DifficultyID int    `json:"difficultyId" validate:"required"`
	MuscleIDs    []int  `json:"muscleIds" validate:"required"`
}

type UpdateWorkoutDTO struct {
	Name         string `json:"name" validate:"required"`
	Description  string `json:"description"`
	MuscleIDs    []int  `json:"muscleIds" validate:"required"`
	DifficultyID int    `json:"difficultyID" validate:"required"`
}

File: client/src/core/types/workout.ts
import { Order } from "./common";
import { Difficulty } from "./difficulty";
import { Muscle } from "./muscle";

export type Workout = {
  id: string;
  name: string;
  description: string;
  muscles: Muscle[];
  difficulty: Difficulty;
  favourite: boolean;
  videoUrl?: string;
};

export type WorkoutDTO = {
  id: string;
  name: string;
  description?: string;
  userId: string;
  difficulty: Difficulty;
  muscles: Muscle[];
};

export type CreateWorkoutDTO = Omit<WorkoutDTO, "id" | "difficulty" | "muscles" | "userId"> & {
  difficultyId: number;
  muscleIds: number[];
};

export type UpdateWorkoutDTO = Partial<CreateWorkoutDTO>;

export type QueryWorkoutsDTO = Partial<{
  filter: FilterWorkoutsDTO;
  order: OrderWorkoutsDTO;
}>;

export type FilterWorkoutsDTO = Partial<{
  name: string;
  muscleIds: number[];
  difficultyId: number;
}>;

export type OrderWorkoutsDTO = Partial<{
  name: Order;
  createdAt: Order;
}>;
